pulsar/internal/connection.go

```diff
@@ -331,10 +331,10 @@ func (c *connection) waitUntilReady() error {
 }
 
 func (c *connection) failLeftRequestsWhenClose() {
+	close(c.incomingRequestsCh)
```

Comment:
I don't think it is safe to close incomingRequestsCh here.
Won't it be panic on producer side of this channel?

Reply:
You're right, it may cause panic in `SendRequest`/`SendRequestNoWait` since channel state check and `incomingRequestsCh` is both not atomic.
As a dirty workaround i've added double check before write to `incomingRequestsCh` by ensuring `closeCh` not closed yet - see next commit

Reply:
Sadly, the workaround won't works. You can test it with the following code with recent golang version. It will panic very quickly.

```golang
func main() {
	c1 := make(chan struct{})
	c2 := make(chan struct{})
	close(c1)
	close(c2)
	for {
		select {
		case <-c1:
		case c2 <- struct{}{}:
		}
	}
}

Reply:
My bad, completely miss that `case`s in `select` has undefined order of execution in general. At the moment I don't have any idea how to resolve data race here but wrap `SendRequest`/`SendRequestNoWait` with mutex, but I don't think it's appropriate solution.

Will reopen PR if find a suitable solution.